RFC-0005-structured-kernel-definitions.md

```diff
@@ -146,127 +145,208 @@ matched with their functional/inplace/out variants):
-together.  Probably the most minimal syntax change here is to just
-introduce funcs which lets you bundle together:
+// Parent class for all code-generated meta:: classes
+struct MetaBase {
+  // TODO: Maybe some of these should be optional, but in many cases they
+  // be implicitly made optional by passing an empty list
+  virtual void set_output(int64_t output_idx, IntArrayRef size, IntArrayRef strides, TensorOptions options, DimnameList names) = 0;
 
-```
-- funcs:
-    - upsample_nearest1d(Tensor self, int[1] output_size, float? scales=None) -> Tensor
-    - upsample_nearest1d.out(Tensor self, int[1] output_size, float? scales=None, *, Tensor(a!) out) -> Tensor(a!)
-  dispatch:
-    CPU: upsample_nearest1d_structured_cpu
-    CUDA: upsample_nearest1d_structured_cuda
-```
+  // Convenience helpers
+  void set_output(IntArrayRef size, TensorOptions options) {
+    set_output(0, size, {}, options, {});
+  }
+};
 
-This complicates ``native_functions.yaml`` ingestion for people who
-don't care about grouping.
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ //
+// Code generated per operator
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ //
 
-## Functions to define
+namespace meta {
 
-Structured definitions require a different set of functions to be
-written to implement the operator.  In this particular case, the
-functions you have to implement are:
+struct upsample_nearest1d : public MetaBase {
+  upsample_nearest1d(const Tensor& self, IntArrayRef output_size, optional<double> scales); // user defined
+};
 
-```
-namespace meta {
- TensorMeta upsample_nearest1d(
- const Tensor& self, IntArrayRef output_size, optional<double> scales);
- }
- 
- namespace native {
- // Precondition: out is an allocated and appropriately sized tensor;
- // all shape checks have passed, device guards have been set,
-// version counter bumps are all handled, etc...
- void upsample_nearest1d_structured_cpu(
-const TensorMeta& out_meta, const Tensor& out, const Tensor& self, IntArrayRef output_size, optional<double> scales);
- void upsample_nearest1d_structured_cuda(
-const TensorMeta& out_meta,  const Tensor& out, const Tensor& self, IntArrayRef output_size, optional<double> scales);
- }
-```
+} // namespace meta
 
-The code generator then generates the boilerplate code to put these
-functions together:
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ //
+// Code generated per dispatch table entry for operator
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ //
 
-```
 namespace native {
 
+struct upsample_nearest1d_cuda : public meta::upsample_nearest1d {
+  void impl(const Tensor& self, IntArrayRef output_size, optional<double> scales); // user defined
+};
+
+// functional implementation
+
+// NB: set_output could be devirtualized with CRTP, but for now we don't do this
+struct upsample_nearest1d_cuda_functional final : public upsample_nearest1d_cuda {
```

Comment:
Missing constructor?

Reply:
It turns out I can't have meta functions be constructors, because they want to call virtual methods but virtual method calls are not valid in constructors. So I am moving theme to be their own method.